homeassistant/components/portainer/config_flow.py

```diff
@@ -0,0 +1,122 @@
+
+_LOGGER = logging.getLogger(__name__)
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+        vol.Required(CONF_API_KEY): str,
+        vol.Optional(CONF_VERIFY_SSL, default=True): bool,
+    }
+)
+
+
+async def _validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect."""
+
+    client = Portainer(
+        api_url=data[CONF_HOST],
+        api_key=data[CONF_API_KEY],
+        session=async_get_clientsession(hass),
+    )
+
+    try:
+        endpoints = await client.get_endpoints()
+    except PortainerAuthenticationError:
+        raise InvalidAuth from None
+    except PortainerConnectionError as err:
+        raise CannotConnect from err
+    except PortainerTimeoutError as err:
+        raise PortainerTimeout from err
+
+    _LOGGER.debug("Connected to Portainer API: %s", data[CONF_HOST])
+
+    for endpoint in endpoints:
+        assert endpoint.id
+        containers = await client.get_containers(endpoint.id)
+        _LOGGER.debug(
+            "Found %d containers on endpoint %s", len(containers), endpoint.name
+        )
+
+    return {"title": data[CONF_HOST]}
+
+
+class PortainerConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Portainer."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                api = await _validate_input(self.hass, user_input)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except PortainerTimeout:
+                errors["base"] = "timeout_connect"
+            except Exception:
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                await self.async_set_unique_id(user_input[CONF_API_KEY])
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(title=api["title"], data=user_input)
+
+        return self.async_show_form(
+            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+        )
+
+    async def async_step_reauth(
+        self, entry_data: Mapping[str, Any]
+    ) -> ConfigFlowResult:
+        """Perform reauth upon an API authentication error."""
+
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Dialog that informs the user that reauth is required."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="reauth_confirm",
+            )
+
+        return await self.async_step_user()
```

Comment:
Note that with reauth, we should only ask the user the api key as the host and ssl stays the same. So Let's split this off for now and implement later